DirectX11/HackerDXGI.cpp

```diff
@@ -813,7 +909,73 @@ STDMETHODIMP HackerDXGIFactory::CreateSwapChain(THIS_
 		return E_FAIL;
 	}
 
-	HRESULT hr = mOrigFactory->CreateSwapChain(origDevice, pDesc, ppSwapChain);
+	HRESULT hr;
+
+	HackerDXGISwapChain *swapchainWrap = nullptr;
+	bool set_fullscreen_required = false;
+
+	if (G->SCREEN_UPSCALING > 2 || G->SCREEN_WIDTH <= 0 || G->SCREEN_HEIGHT <= 0 || G->UPSCALE_MODE >=2)
+	{
+		// TODO: some default parameters for upscaling could be added here
+		// TODO: what if user provides wrong screen width or wrong screen height
+		// TODO: maybe enumeration of modes and resolutions should be applied here
+		LogInfo("At least one of provided upscaling paramters is invalid!");
+		LogInfo("Please check the d3d11.ini file!");
+		LogInfo("Upscaling is disabled!");
+		G->SCREEN_UPSCALING = 0;
+	}
+	
+	if (G->SCREEN_UPSCALING == 0) // no upscaling at all use the usual "old" path
+	{
+		ForceDisplayParams(pDesc);
+		hr = mOrigFactory->CreateSwapChain(origDevice, pDesc, ppSwapChain);
+		swapchainWrap = new HackerDXGISwapChain(*ppSwapChain, hackerDevice, hackerDevice->GetHackerContext());
+	}
+	else
+	{	
+		set_fullscreen_required = !pDesc->Windowed; // to create two swap chains in row we need the swap chains to be in windowed mode
+		pDesc->Windowed = 1; // so store has to be stored to call setFullscreen after and changed to windowed
+							 // NOTE: no ForceDisplayResolution required here
+		InstallSetWindowPosHook(); // This hook is very important in case of upscaling
+								   //TODO: what about the hook and the warning in forceDisplayResolution (Testing required)
+								   //SetWindowPos(pDesc->OutputWindow, nullptr, 0, 0, 0, 0, SWP_HIDEWINDOW);
+		DXGI_SWAP_CHAIN_DESC olddesc = *pDesc; // need old description to create fake swap chain (fake texture)
+
+		// TODO: testing testing testing
+		// not sure if the upscaling will work that way
+		// maybe i should better disable upscaling if 3dv direct mode is set
+		if (G->gForceStereo == 2)
```

Comment:
No, no game presently uses the Direct Mode that I added.  It's currently specific to a Unity plugin that sgsrules is creating.  First game we expect to see use it is Inside. 

The TombRaider uses their built-in 3D.  Similar to GTA5 and Trine.  Those probably actually do use the NVidia Direct Mode in the driver, but won't need this code because they already handle it.  

Pretty sure the SBS shader will work on TombRaider as it stands, no normal 3Dmigoto code, just the SBS shader at the end.  Seems like Upscaling should also be fine.